### ISSUE TITLE
loadWorkbook removing formatting

### ISSUE THREAD

**Issue by mdorchuckuva** [ycphs/openxlsx#207]
When using loadWorkbook, formatting that has been applied by a user in excel directly is being stripped.  

With the attached excel file in my working directory, running the following code opens a file with cell C4 in black, unbolded text instead of the red, bold text as originally in the file.  The formatting remains stripped if I save the wb object back out and then open from file explorer. 

```
library(openxlsx)
wb = loadWorkbook("testopenxlsx.xlsx")
openXL(WB)
```

System information
Windows 10 Enterprise 10.0.19042
Have tested on R 4.0.3 and R 4.1.0, in both cases the issue persists
Openxlsx version 4.2.4
Office 365 version 2102
[testopenxlsx.xlsx](https://github.com/ycphs/openxlsx/files/6711990/testopenxlsx.xlsx)

**Comment by mdorchuckuva** [comment]
I found that the issue no longer occurs if I back up my openxlsx version to 4.0.0, in case that helps narrow it down.

**Comment by ycphs** [comment]
In your example is you open a different workbook than the one you loaded

```
library(openxlsx) 
wb = loadWorkbook("testopenxlsx.xlsx")
openXL(WB)
```

Please try the following:


```
library(openxlsx) 
wb = loadWorkbook("testopenxlsx.xlsx")
openXL(wb)
```

**Comment by mdorchuckuva** [comment]
Sorry the wb got capitalized when I pasted over for some reason.  I was running the same workbook.  see screenshot below and thanks so much for taking a look!

Top workbook is what opens from OpenXL, bottom workbook is the workbook that was originally saved.  

![openxlsx_screenshot](https://user-images.githubusercontent.com/27371304/123335751-e8fdb200-d512-11eb-8faf-6514e77760b1.png)

**Comment by JanMarvin** [comment]
I've seen similar issues before. What I have found is that the style order is changed by `openxlsx`. We always add our default style to `styles.xml`, which is picked here and all the other styles are re-sorted and do no longer match with the style ids in the worksheet. Iirc the styles file has no style id field and they are picked in the order of appearance. Kinda the same issue as #81.

**Comment by Seviks** [comment]
I can confirm this issue on 4.2.4. When I `loadWorkbook`, do my changes (or not) an then `saveWorkbook`, the formatting is gone. 

Formats persisted in 4.2.3. After roll back to 4.2.3 the issue is gone.

Tested on:
openxlsx 4.2.4
Windows 10
R 4.1.0
Office 2019

**Comment by patrick-gerland** [comment]
Indeed, by upgrading to version 4.2.4 this new issue happens. Rolling back to the previous version fixes the problem.
I spent most of 1 day troubleshooting what happened to all my workbooks (>200 with multiple worksheets).

Just loading and saving back the file causes the problem with loosing the existing settings for the default column headers.
		wb <- loadWorkbook(input.file)
		saveWorkbook(wb, input.file, overwrite = T)

The initial wb simply contains 1 worksheet like
addWorksheet(wb, sheet = "myInput")
writeDataTable(wb, sheet = "myInput", x = myInput, colNames = TRUE, rowNames = FALSE, headerStyle = hs1, tableStyle = "TableStyleLight2", withFilter = FALSE, bandedRows = TRUE)

with 
hs1 <- createStyle(fgFill = "#4F81BD", halign = "CENTER", textDecoration = "Bold", border = "Bottom", fontColour = "white")

with the newer version fontColour setting gets lost to some dark default instead of white.

**Comment by derekunderwood** [comment]
Similar issue here although format is being altered, not removed entirely.  Seemed appropriate to post here but please move if separate issue.

Cells formatted single accounting underline using this option:
![image](https://user-images.githubusercontent.com/56615756/125679761-32f88b06-c75f-4617-914b-c722ad291e44.png)

Appear to load with normal underline formatting:
![openxlsx](https://user-images.githubusercontent.com/56615756/125680052-fa761613-1b38-41a5-86f6-4cbe0fec7ef5.PNG)

Tested on:
openxlsx 4.2.3
Windows 10
R 3.6.3
Office 2016

[underline_test.xlsx](https://github.com/ycphs/openxlsx/files/6818492/underline_test.xlsx)

**Comment by JanMarvin** [comment]
A potential fix is in #237 could any of you test this please?

**Comment by patrick-gerland** [comment]
Thanks lot. I just installed the latest development version
packageVersion("openxlsx")
 ‘4.2.4.9000’

Unfortunately the problem remains.

**Comment by mdorchuckuva** [comment]
Thanks so much @JanMarvin.  Sorry for the delayed response here.   

I built #237 on Windows.  Results are shown in the screenshot.  Left book is the input, right book is what opens from openxlsx::openXL.  We keep font changes, bolding and underlines.  We still lose merged cell, color changes, and strikethrough.  Input workbook also attached for your reference.

Please let me know if I can do anything to help that involves R (do not know cpp, sorry). 

![image](https://user-images.githubusercontent.com/27371304/128426881-b3711a1e-9243-4cf6-8c22-dd4a8c1e71ec.png)
[testopenxlsx.xlsx](https://github.com/ycphs/openxlsx/files/6941882/testopenxlsx.xlsx

**Comment by JanMarvin** [comment]
Thanks for testing (the pull request isn't yet merged with the master branch, I wanted to get feedback first).

@mdorchuckuva do you remember if these used to work with prior releases or did you list them for being important?

**Comment by mdorchuckuva** [comment]
@JanMarvin thanks! 

Yes we had a report that contained some of these features (particularly merged cells) that broke after an update to openxlsx. 

I can confirm the test workbook works correctly across all items except strikethrough if I roll back to 4.0.0. 

![image](https://user-images.githubusercontent.com/27371304/128433391-9007c607-8302-478b-a007-4477021c6fc5.png)

**Comment by mdorchuckuva** [comment]
Edited above comment for clarity.

**Comment by ycphs** [comment]
The commit for version 4.0.0 is e99f707c2134b936a23e0e26327f354775f2e4d6
At the previous repo (awalker89/openxlsx) at 5 Jan 2017.

It will be tricky to find the cause, but it is good to know when it worked the last time.

**Comment by JanMarvin** [comment]
I assume it might be a more recent regression. I'll try to look into it.

**Comment by JanMarvin** [comment]
![Screenshot_2021-08-06_21-46-32](https://user-images.githubusercontent.com/1645626/128563872-dcadff8c-a376-4f55-9dfc-120b24e76c7c.png)

Hi @mdorchuckuva, I assume I've fixed it. Please have another look. I've even added `<strike/>` (because it was easy). Both files look identical for me. Could you please have a look?

**Comment by mdorchuckuva** [comment]
@JanMarvin you looking at PR #237?  I checked that PR out - can build on windows but am getting a fatal error in RStudio w/ no message or other info when I try to execute `loadworkbook`.  Any advice?  Am I in the right PR?  

Initial (uninformed) guess is that it may be related to which version of rcpp I'm using - I'm on 1.0.7.

**Comment by JanMarvin** [comment]
yeah, that's the one. not sure what's going on. it works on the ci (all green on GitHub actions). No clue why it should fail locally ~~and I assume it's related to your setup~~. 😞 

Something like this should work
```R
library(remotes)
install_github("ycphs/openxlsx", ref = github_pull(237))

library(openxlsx)
wb <- loadWorkbook(file = system.file("extdata", "loadExample.xlsx", package = "openxlsx"))

```

Edit: Okay, I can confirm the crash on windows 10.

**Comment by JanMarvin** [comment]
I've rebased the pull request on master and can now run it on windows (maybe the issue was caused by some unrelated pending fix or an RStudio bug?). I freshly installed R and Rstudio and a bunch of packages (usually use my Windows only for gaming 😄 ).

```R
> sessionInfo()
R version 4.1.0 (2021-05-18)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19042)

Matrix products: default

locale:
[1] LC_COLLATE=German_Germany.1252  LC_CTYPE=German_Germany.1252    LC_MONETARY=German_Germany.1252
[4] LC_NUMERIC=C                    LC_TIME=German_Germany.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] openxlsx_4.2.4.9000

loaded via a namespace (and not attached):
[1] compiler_4.1.0 tools_4.1.0    Rcpp_1.0.7     stringi_1.7.3  zip_2.2.0
```

**Comment by JanMarvin** [comment]
Should be fixed in https://github.com/ycphs/openxlsx/commit/0fc1c08ceebfa6a9fd50cd54c4bb55459b2900a5

**Comment by mdorchuckuva** [comment]
@JanMarvin sorry didn't see this over the weekend.  FYI I can execute the prior test code successfully after installing via remotes, w/out updating any of the referenced packages or re-installing R/Rstudio/packages.  So it seems to me that the fail issue was isolated to building that PR via Rstudio on Win10, not should occur if installed normally once pushed to CRAN.  

Thanks so much for your help w/ this, really appreciate it!

**Comment by JanMarvin** [comment]
Thanks for the feedback! I assume that a new Cran release will follow soon

**Comment by neros** [comment]
> Thanks for the feedback! I assume that a new Cran release will follow soon

Thank you @JanMarvin and others for the fix.  Is there a timeline for this fix on Cran?  Having upgraded to 4.2.4, existing code breaks that uses styles (filtering data using FontColour format, specifically). This is an important fix for existing code base within the package community.

**Comment by JanMarvin** [comment]
Hi @neros , I'm not involved in the CRAN releasing process. That's @ycphs part, but he was involved with personal matters. For now the current master is what I consider the next release, therefore if you need it, you can build from there. Otherwise a new release will follow soon™.

**Comment by mendy-friedman** [comment]
I downloaded the latest master branch. Most of the formatting issues are now fixed, but there are still a few formatting pieces that are not carried over.  Can you please look into this before the next release?

- Tab formatting: If a tab in the original workbook has a color instead of the default gray, this color is lost upon loading and then saving the workbook.  Additionally, hidden tabs become unhidden upon loading and saving the workbook.
- Form controls (action buttons): some of these resize upon saving the workbook
- Hidden columns: If you hide a single column, it remains hidden upon loading and saving the workbook.  But if you hide all columns from column X to the end of the workbook, those columns unhide after saving.

Some screenshots are below.  I can't share the full workbook for confidentiality reasons, but it should be easy to create a reproducible example based on these comments.

**Original (tabs are colored, some tabs are hidden, buttons are a certain size, and some columns are hidden):**
![image](https://user-images.githubusercontent.com/93217100/138931404-13b9b194-b6ca-4b8b-832a-9514b43319de.png)
![image](https://user-images.githubusercontent.com/93217100/138931817-edc1640d-dce8-41b6-a0dd-3cc288df6526.png)
![image](https://user-images.githubusercontent.com/93217100/138932118-0331b000-f644-420d-84d5-510f32cb2b77.png)

**After Saving (tabs are all gray, and all are unhidden.  Columns unhide, and buttons are wider):**
![image](https://user-images.githubusercontent.com/93217100/138931459-44452382-f2e3-49b9-a31b-45294d06f9ac.png)
![image](https://user-images.githubusercontent.com/93217100/138931881-10900d70-4997-4786-996f-2d72a3b83e11.png)
![image](https://user-images.githubusercontent.com/93217100/138932158-9cfcf588-d901-4647-af6d-91ee5ea9012c.png)

**Comment by JanMarvin** [comment]
Hi @mendy-friedman , I currently don't have time to look into this and personally think these can be fixed in follow up releases. It might be a good idea to open additional issues for these and just out of curiosity, did any of these work in previous releases? For instance the hidden/unhidden part seems like it never worked and the button size is afaik unhandled too. Also it would be helpful if you could provide pre and post xlsx-files.

**Comment by JanMarvin** [comment]
Late night test. I created a simple  workbook with colored tabs with MS 365 in the web editor. Downloaded the xlsx file. Ran the following lines and the resulting `test.xlsx` sheet had colored tabs with LibreOffice and MS 365. Meaning I couldn't reproduce the issue. Same thing happend with hidden columns. If I hid all from `C:XFD` in MS 365 and ran the lines below, everything looked the same (with expected errors on LibreOffice, but fine in MS 365). Therefore if something isn't working please open new issues and provide steps and files to reproduce the issues.

```R
library(openxlsx)
fl <- "~/ColorTabs.xlsx"

wb <- loadWorkbook(fl)
saveWorkbook(wb, "/tmp/test.xlsx")
```

Edit: hidden sheets unhide. That's unfortunate, but nothing to serious.

**Comment by mendy-friedman** [comment]
appreciate the quick reply - I'll try to create a reproducible example that I can share.  I'll reply to the separate issue that you opened.

**Comment by mendy-friedman** [comment]
In my actual use case, the worksheet tabs do lose their colors, but when I create a (simpler) reproducible example, the tabs maintain their colors.  Wondering if one of the steps that I do before saving (writing new data to the workbook, etc.) is impacting this.  I'll keep researching.

The issue with attributes that apply to an entire column or row being deleted is still happening unfortunately.  I will open a new issue for this and share an example.  Thank you for your help.

**Comment by mdorchuckuva** [comment]
@mendy-friedman not a dev on this project, but a frequent user.  I have experienced what you're experiencing (tabs losing colors, some columns/rows losing attributes) if there is an xml error w/ the workbook generally.  when you do saveworkbook on the one you're seeing issues w/ in R are you getting any warning messages back?

do you have comments in the workbook (the new office 365 style)?

**Comment by mendy-friedman** [comment]
@mdorchuckuva - I'm not getting any warning messages back.  No office 365 comments in the workbook, although it does include some other attributes such as some vba code, a user form, etc.

@JanMarvin - I was able to create a reproducible example that doesn't include company confidential information.  When I read in the first file and save it back out, I get the second file (the tab color of the first tab is stripped).  It's an xlsm file, so I can't upload the file the normal way to this chat (only .xls and .xlsx are supported), but I was able to put the input and output files into a .zip file and uploaded that below.  Let me know if you have troubles accessing the files, or if you'd like me to open this as a separate issue.
[testfiles.zip](https://github.com/ycphs/openxlsx/files/7427381/Desktop.zip)

**Comment by JanMarvin** [comment]
Hi @mendy-friedman , long time no see. I opened the file with `openxlsx2` which should be able to handle your file just fine. I ran the following code and the files look identical, but the sheets are not colored. If you want to give it a try, see the code below. I was using the current `openxlsx2` main branch.

```R
# remotes::install_github ...
library(openxlsx2)
library(curl) # for the download

### prepare
tmp <- tempdir()
input_zip <- paste0(tmp, "/Desktop.zip")

input <- paste0(tmp, "/test2.xlsm")
output <- paste0(tmp, "/test_openxlsx.xlsm")

### get file
curl::curl_download(
  "https://github.com/ycphs/openxlsx/files/7427381/Desktop.zip",
  destfile = input_zip)

unzip(input_zip, exdir = tmp)
dir(tmp)

### actual file work
#  load the workbook
wb <- wb_load(input)
# wb$open()
wb$save(output)

### compare
xl_open(input)
xl_open(output)
```

**Comment by mendy-friedman** [comment]
Thanks @JanMarvin - I'll check it out.  I didn't even know that `openxlsx2` existed (we've still been using `openxlsx` for our existing pipelines). I'll have to explore it.  Reading the vignette the enhancements look exciting.  Thanks for the heads up!

**Comment by JanMarvin** [comment]
Development is still quite active. Our first CRAN release 0.3 is available since last month and a bugfix 0.3.1 release shall follow this month. For all openxlsx2 questions please use the other bug tracker and discussion boards.
